pkg/didcomm/protocol/messagepickup/service.go

```diff
@@ -0,0 +1,612 @@
+	}
+
+	return nil
+}
+
+// StatusRequest request a status message
+func (s *Service) StatusRequest(connectionID string) (*Status, error) {
+	// get the connection record for the ID to fetch DID information
+	conn, err := s.getConnection(connectionID)
+	if err != nil {
+		return nil, err
+	}
+
+	// generate message ID
+	msgID := uuid.New().String()
+
+	// register chan for callback processing
+	statusCh := make(chan Status)
+	s.setStatusCh(msgID, statusCh)
+
+	defer s.setStatusCh(msgID, nil)
+
+	// create request message
+	req := &StatusRequest{
+		Type: StatusRequestMsgType,
+		ID:   msgID,
+		Thread: &decorator.Thread{
+			PID: uuid.New().String(),
+		},
+	}
+
+	// send message to the router
+	if err := s.outbound.SendToDID(req, conn.MyDID, conn.TheirDID); err != nil {
+		return nil, fmt.Errorf("send route request: %w", err)
+	}
+
+	// callback processing (to make this function look like a sync function)
+	var sts *Status
+	select {
+	case s := <-statusCh:
+		sts = &s
+	// TODO https://github.com/hyperledger/aries-framework-go/issues/1134 configure this timeout at decorator level
+	case <-time.After(updateTimeout):
+		return nil, errors.New("timeout waiting for status request")
+	}
+
+	return sts, nil
+}
+
+// BatchPickup a request to have multiple waiting messages sent inside a batch message
+func (s *Service) BatchPickup(connectionID string, size int) (int, error) {
+	// get the connection record for the ID to fetch DID information
+	conn, err := s.getConnection(connectionID)
+	if err != nil {
+		return -1, err
+	}
+
+	// generate message ID
+	msgID := uuid.New().String()
+
+	// register chan for callback processing
+	batchCh := make(chan Batch)
+	s.setBatchCh(msgID, batchCh)
+
+	defer s.setBatchCh(msgID, nil)
+
+	// create request message
+	req := &BatchPickup{
+		Type:      BatchPickupMsgType,
+		ID:        msgID,
+		BatchSize: size,
+	}
+
+	// send message to the router
+	if err := s.outbound.SendToDID(req, conn.MyDID, conn.TheirDID); err != nil {
+		return -1, fmt.Errorf("send route request: %w", err)
+	}
+
+	// callback processing (to make this function look like a sync function)
+	var processed int
+	select {
+	case batchResp := <-batchCh:
+		for _, msg := range batchResp.Messages {
+			err := s.handle(msg)
+			if err != nil {
+				logger.Errorf("error handling batch message %s: %v", msg.ID, err)
+				continue
+			}
+			processed++
+		}
+	// TODO https://github.com/hyperledger/aries-framework-go/issues/1134 configure this timeout at decorator level
+	case <-time.After(updateTimeout):
+		return -1, errors.New("timeout waiting for batch")
+	}
+
+	return processed, nil
+}
+
+func (s *Service) getConnection(routerConnID string) (*connection.Record, error) {
+	conn, err := s.connectionLookup.GetConnectionRecord(routerConnID)
+	if err != nil {
+		if errors.Is(err, storage.ErrDataNotFound) {
+			return nil, ErrConnectionNotFound
+		}
+
+		return nil, fmt.Errorf("fetch connection record from store : %w", err)
+	}
+
+	return conn, nil
+}
+
```

Comment:
Can be simplified 
```
func (s *Service) getConnection(routerConnID string) (*connection.Record, error) {
	conn, err := s.connectionLookup.GetConnectionRecord(routerConnID)
	if errors.Is(err, storage.ErrDataNotFound) {
		return nil, ErrConnectionNotFound
	}

	return conn, err
}
``

Reply:
I think it's a good idea to wrap the error that is not `storage.ErrDataNotFound`.

Reply:
@soluchok / @llorllale can one of you reslove this, I am easy as to which to adopt.

Reply:
it is ok, you can leave as it is